#!/usr/bin/env python3
"""
–¢–µ—Å—Ç–æ–≤—ã–π —Å–∫—Ä–∏–ø—Ç –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –Ω–æ–≤—ã—Ö –º–µ—Ç–æ–¥–æ–≤ API –∏—Å—Ç–æ—Ä–∏–∏ —É—Å—Ç—Ä–æ–π—Å—Ç–≤
"""

import asyncio
import sys
from grpclib.client import Channel
from marznode.service.service_grpc import MarzServiceStub
from marznode.service.service_pb2 import Empty, UserDevicesRequest


async def test_fetch_user_devices(stub, uid: int, active_only: bool = False):
    """–¢–µ—Å—Ç –ø–æ–ª—É—á–µ–Ω–∏—è —É—Å—Ç—Ä–æ–π—Å—Ç–≤ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è"""
    print(f"\n=== Testing FetchUserDevices for user {uid} (active_only={active_only}) ===")
    
    try:
        response = await stub.FetchUserDevices(
            UserDevicesRequest(uid=uid, active_only=active_only)
        )
        
        print(f"User ID: {response.uid}")
        print(f"Total devices: {len(response.devices)}")
        
        for idx, device in enumerate(response.devices, 1):
            print(f"\nDevice {idx}:")
            print(f"  Remote IP: {device.remote_ip}")
            print(f"  Client Name: {device.client_name}")
            print(f"  User Agent: {device.user_agent}")
            print(f"  Protocol: {device.protocol}")
            print(f"  TLS Fingerprint: {device.tls_fingerprint}")
            print(f"  First Seen: {device.first_seen}")
            print(f"  Last Seen: {device.last_seen}")
            print(f"  Total Usage: {device.total_usage} bytes")
            print(f"  Uplink: {device.uplink} bytes")
            print(f"  Downlink: {device.downlink} bytes")
            print(f"  Is Active: {device.is_active}")
        
        return response
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
        return None


async def test_fetch_all_devices(stub):
    """–¢–µ—Å—Ç –ø–æ–ª—É—á–µ–Ω–∏—è –≤—Å–µ—Ö —É—Å—Ç—Ä–æ–π—Å—Ç–≤"""
    print("\n=== Testing FetchAllDevices ===")
    
    try:
        response = await stub.FetchAllDevices(Empty())
        
        print(f"Total users with devices: {len(response.users)}")
        total_devices = sum(len(user.devices) for user in response.users)
        print(f"Total devices across all users: {total_devices}")
        
        for user_history in response.users:
            print(f"\nUser ID: {user_history.uid}")
            print(f"  Devices: {len(user_history.devices)}")
            
            for idx, device in enumerate(user_history.devices, 1):
                status = "üü¢ Active" if device.is_active else "üî¥ Inactive"
                print(f"    {idx}. {device.remote_ip} ({device.client_name}) - {status}")
                print(f"       Usage: {device.total_usage} bytes")
        
        return response
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
        return None


async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    host = sys.argv[1] if len(sys.argv) > 1 else "localhost"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 53042
    
    print(f"Connecting to {host}:{port}")
    
    # –°–æ–∑–¥–∞—ë–º insecure —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–µ (–¥–ª—è —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è)
    # –í –ø—Ä–æ–¥–∞–∫—à–µ–Ω–µ –Ω—É–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å SSL
    channel = Channel(host, port)
    stub = MarzServiceStub(channel)
    
    try:
        # –¢–µ—Å—Ç 1: –ü–æ–ª—É—á–∏—Ç—å –≤—Å–µ —É—Å—Ç—Ä–æ–π—Å—Ç–≤–∞
        await test_fetch_all_devices(stub)
        
        # –¢–µ—Å—Ç 2: –ü–æ–ª—É—á–∏—Ç—å —É—Å—Ç—Ä–æ–π—Å—Ç–≤–∞ –∫–æ–Ω–∫—Ä–µ—Ç–Ω–æ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
        # (–∑–∞–º–µ–Ω–∏—Ç–µ 1 –Ω–∞ —Ä–µ–∞–ª—å–Ω—ã–π UID –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è)
        if len(sys.argv) > 3:
            uid = int(sys.argv[3])
            await test_fetch_user_devices(stub, uid, active_only=False)
            await test_fetch_user_devices(stub, uid, active_only=True)
        else:
            print("\n\n–¢–∏–ø: python test_devices_api.py [host] [port] [user_id]")
            print("–ü—Ä–∏–º–µ—Ä: python test_devices_api.py localhost 53042 1")
    
    finally:
        channel.close()


if __name__ == "__main__":
    asyncio.run(main())

